crates/tx5/src/proto.rs

```diff
@@ -0,0 +1,580 @@
+//! Types associated with the Tx5 protocol.
+
+use crate::BackBuf;
+use tx5_core::{Error, Result};
+
+/// Tx5 protocol message payload max size.
+pub(crate) const MAX_PAYLOAD: u32 = 0b00011111111111111111111111111111;
+
+/// Protocol version 1 header.
+pub(crate) const PROTO_VER_1: ProtoHeader =
+    ProtoHeader::Version(1, [b't', b'x', b'5']);
+
+/// 4 byte Tx5 protocol header.
+///
+/// The initial 3 bits representing values 0, 1, and 7 are reserved.
+/// Decoders should error on receiving these values.
+#[derive(Debug, PartialEq)]
+pub(crate) enum ProtoHeader {
+    /// This is a protocol version handshake.
+    /// `3 bits == 2`
+    /// `5 bits == version number (currently 1)`
+    /// `24 bytes == as bytes b"tx5"`
+    Version(u8, [u8; 3]),
+
+    /// The remainder of this message represents the entirety of a message.
+    /// `3 bits == 3`
+    /// `29 bits == the message size == full chunk size - 4`
+    CompleteMessage(u32),
+
+    /// The remainder of this message is a chunk of a multi-part message.
+    /// `3 bits == 4`
+    /// `29 bits == byte count included in this message chunk`
+    MultipartMessage(u32),
+
+    /// This is a request for a permit to send a message of size larger than
+    /// a single chunk.
+    /// `3 bits == 5`
+    /// `29 bits == the size of the message`
+    PermitRequest(u32),
+
+    /// This is an authorization to proceed with sending payload chunks.
+    /// `3 bits == 6`
+    /// `29 bits == the size of the message`
+    PermitGrant(u32),
+}
+
+impl ProtoHeader {
+    /// Decode 4 bytes into a Tx5 protocol header.
+    pub fn decode(a: u8, b: u8, c: u8, d: u8) -> Result<Self> {
+        use bit_field::BitField;
+        let r = u32::from_be_bytes([a, b, c, d]);
+        match r.get_bits(29..) {
+            2 => Ok(Self::Version(
+                r.get_bits(24..29) as u8,
+                [
+                    r.get_bits(16..24) as u8,
+                    r.get_bits(8..16) as u8,
+                    r.get_bits(0..8) as u8,
+                ],
+            )),
+            3 => Ok(Self::CompleteMessage(r.get_bits(..29))),
+            4 => Ok(Self::MultipartMessage(r.get_bits(..29))),
+            5 => Ok(Self::PermitRequest(r.get_bits(..29))),
+            6 => Ok(Self::PermitGrant(r.get_bits(..29))),
+            _ => Err(Error::id("ReservedHeaderBits")),
+        }
+    }
+
+    /// Encode a Tx5 protocol header into canonical 4 bytes.
+    pub fn encode(&self) -> Result<(u8, u8, u8, u8)> {
```

Comment:
This seems easier to work with but maybe it's just a matter of preference

```suggestion
    pub fn encode(&self) -> Result<[u8; 4]> {
```

Reply:
For symmetry with the decode side, I'd prefer to keep them separate... You'd *think* you'd want `decode(data: [u8; 4])` BUT, that means you have to:

```rust
fn my_fn(in: Vec<u8>) {
    assert!(in.len() >= 4);
    decode([in[0], in[1], in[2], in[3]])
```

because rust doesn't have any straight forward way to use a Vec or `&[u8]` as a `[u8; 4]`...

Reply:
Well there is try_into which is pretty straightforward: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3fb3a2f346241d2012a6d1a033385017